spec/Overview.html

```diff
@@ -10168,111 +10139,1879 @@ <h4>Operations</h4>
+            </tbody>
+          </table>
+        </section>
+
+        <section id="x25519-operations">
+          <h4>Operations</h4>
+          <dl>
+            <dt>Derive Bits</dt>
+            <dd>
+              <ol>
+                <li>
+                  <p>
+                    If the {{CryptoKey/[[type]]}} internal slot of
+                    |key| is not {{KeyType/"private"}}, then [= exception/throw =] an {{InvalidAccessError}}.
+                  </p>
+                </li>
+                <li>
+                  <p>
+                    Let |publicKey| be the
+                    {{EcdhKeyDeriveParams/public}} member of
+                    |normalizedAlgorithm|.
+                  </p>
+                </li>
+                <li>
+                  <p>
+                    If the {{CryptoKey/[[type]]}} internal slot of
+                    |publicKey| is not "`public`", then [= exception/throw =] an {{InvalidAccessError}}.
+                  </p>
+                </li>
+                <li>
+                  <p>
+                    If the {{KeyAlgorithm/name}} attribute of
+                    the {{CryptoKey/[[algorithm]]}} internal slot of
+                    |publicKey| is not equal to the {{KeyAlgorithm/name}} property of the {{CryptoKey/[[algorithm]]}} internal slot of
+                    |key|, then [= exception/throw =] an {{InvalidAccessError}}.
+                  </p>
+                </li>
+                <li>
+                  <p>
+                    Let |secret| be the result of performing the X25519 function specified in
+                    [[RFC7748]] Section 5 with |key| as the X25519 private key |k|
+                    and the X25519 public key represented by the {{CryptoKey/[[handle]]}}
+                    internal slot of |publicKey| as the X25519 public key |u|.
+                  </p>
+                </li>
+                <li>
+                  <p>
+                    If |secret| is the all-zero value,
+                    then [= exception/throw =] a {{OperationError}}.
+                    This check must be performed in constant-time, as per [[RFC7748]] Section 6.1.
+                  </p>
+                </li>
+                <li>
+                  <dl class="switch">
+                    <dt>If |length| is null:</dt>
+                    <dd>Return |secret|</dd>
+                    <dt>Otherwise:</dt>
+                    <dd>
+                      <dl class="switch">
+                        <dt>
+                          If the length of |secret| in bits is less than
+                          |length|:
+                        </dt>
+                        <dd>
+                          [= exception/throw =] an
+                          {{OperationError}}.
+                        </dd>
+                        <dt>Otherwise:</dt>
+                        <dd>
+                          Return an [= octet string containing =] the first |length| bits of |secret|.
```

Comment:
I think there are legitimate use cases for shorter secrets, e.g., 128-bit or 192-bit symmetric keys. That being said, instead of using the output of X25519 directly, many protocols pass the output of X25519 through a hash function, often combined with additional data, in which case one truncates the output of the hash function and not the output of X25519.

In any case, the `length` parameter is a positional argument of the public `deriveBits()` function and cannot be removed. Theoretically, the X25519 implementation of the _derive bits operation_ could reject lengths less than 32 bytes, but that seems like a rather artificial restriction and would be inconsistent with ECDH, for example.

Reply:
There has been some discussion on changing the handling of the `length` parameter of `deriveBits` for ECDH, e.g. https://github.com/w3c/webcrypto/pull/345#issuecomment-1682731140 proposed to throw if the requested length is shorter than the full value. I believe Chromium is doing some measurements to see if that would be compatible with existing web applications. If so, we could change it for X25519 as well - but I do think it should be consistent with ECDH, indeed.

Reply:
Thanks for your comments. I get the point regarding consistency with similar existing interfaces.